APE23.rst

```diff
@@ -0,0 +1,251 @@
+
+author: Thomas Robitaille
+
+date-created: 2022 April 20
+
+date-last-revised: 2024 June 10
+
+date-accepted: <replace with accepted date>
+
+type: Process
+
+status: Discussion
+
+
+Abstract
+--------
+
+The original scope of the Astropy project was to develop a single core package,
+and it was decided at the time that the only required dependency should be Numpy
+to maximize portability across platforms and ease of installation. Since then
+however, there has been four instances where a required dependency has been
+added to the core package (packaging, PyERFA, PyYAML, and astropy-iers-data), along with 21
+'optional' dependencies. The extensive use of optional dependencies can cause
+frustration for users who don't realise that some dependencies are not installed
+by default, and there are no clear guidelines for the core package or other
+project packages regarding when dependencies should be considered required or
+optional.
+
+This APE aims to provide a set of guidelines for deciding whether dependencies
+for a package should be optional or required and aspects to consider when adding
+a new dependency at all, with the aim to improve user experience as well as avoid
+extensive discussions amongst developers every time a new dependency is proposed.
+While this APE was originally written with the core package in mind, the guidelines
+are general enough to be applicable to all packages in the astropy ecosystem.
+
+Detailed description
+--------------------
+
+Background
+^^^^^^^^^^
+
+Since the inception of the Astropy project over ten years ago, Python packaging
+has progressed significantly, not least with the availability now of binary
+wheels for packages as well as packaging systems such as conda. At the start of
+the project, getting a scientific Python stack set up (including e.g. NumPy,
+SciPy, and Matplotlib) was not straighforward and required building all these
+packages from source. As a result, the original vision for the project included
+the goal to only have Numpy as a required dependency for the core astropy package.
+
+Since then, there have been four instances where a required dependency has been
+added:
+
+* PyERFA (in 2020): the core astropy package used to include ERFA wrappers, which
+  were then split into a separate package under the name PyERFA. However, this
+  meant adding PyERFA as a required dependency to keep the core astropy working
+  as it has before.
+
+* PyYAML (in 2021): formerly an optional dependency, PyYAML is used in the core
+  astropy package in order to serialize Table information losslessly in several
+  formats. The main reason this was added as a required dependency was because
+  it was not an obvious dependency to users - for instance a user writing a
+  table to ECSV could have no reasonable way of knowing in advance that PyYAML
+  would be required for this.
+
+* packaging (in 2021): used for comparison of versions, to replace the
+  deprecated build-in module ``distutils``.
```

Comment:
"builtin module" has a different meaning (for instance in pypy, `math` is a builtin module, in the sense that it's built into the interpreter and importing it is thus free)
```suggestion
  deprecated standard library module ``distutils``.
```